src/App/Fossa/VSI/Fingerprint.hs

```diff
@@ -0,0 +1,217 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE RoleAnnotations #-}
+
+module App.Fossa.VSI.Fingerprint (
+  fingerprintRaw,
+  fingerprintContentsRaw,
+  fingerprintCommentStripped,
+  fingerprint,
+  Fingerprint,
+  Raw,
+  CommentStripped,
+  Combined (..),
+) where
+
+import Conduit (ConduitT, Void, await, decodeUtf8C, encodeUtf8C, filterC, linesUnboundedC, mapC, runConduitRes, sourceFile, yield, (.|))
+import Control.Algebra (Has)
+import Control.Effect.Diagnostics (Diagnostics, fatalText)
+import Control.Effect.Exception (IOException, Lift, catch)
+import Control.Effect.Lift (sendIO)
+import Crypto.Hash (Digest, HashAlgorithm, SHA256 (..), hashFinalize, hashInit, hashUpdate)
+import Data.Aeson (ToJSON, object, toJSON, (.=))
+import Data.ByteString qualified as B
+import Data.Maybe (fromMaybe)
+import Data.String.Conversion (ToText (..))
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.Extra (breakOnAndRemove)
+import Discovery.Walk (WalkStep (..), walk')
+import Effect.ReadFS (ReadFS, contentIsBinary)
+import Path (Abs, Dir, File, Path, toFilePath)
+
+-- | Fingerprint deterministically idenfies a file and is derived from its content.
+--
+-- The type variable is:
+--
+--   @k@ - Kind, the kind of fingerprint computed.
+--
+-- For ease of implementation, the backing representation of a @Fingerprint@ instance is a @Base16@ encoded @Text@.
+newtype Fingerprint k = Fingerprint Text
+
+type role Fingerprint nominal
+
+-- | Represents a 'Fingerprint' derived from the unmodified content of a file.
+data Raw
+
+-- | Represents a 'Fingerprint' derived from the content of a file with all C-style comments removed.
+data CommentStripped
+
+instance ToText (Fingerprint k) where
+  toText (Fingerprint k) = k
+
+instance ToJSON (Fingerprint k) where
+  toJSON = toJSON . toText
```

Comment:
[nit]: We can probably kill this and just add `deriving (ToJson)` to the newtype declaration.